File: solver/exp.ml
(*
 * DPLL(T) SMT Solver -- Expression Grammar
 *
 * We distinguish between high-level boolean expressions and arithmetic
 * expressions (i.e., expressions in a special 'theory'). Boolean
 * expressions may contain arithmetic expressions but not the other way
 * around.
 *
 * This means that we have both propositional variables (e.g., "p = true")
 * and arithmetic variables (e.g., "x = 5"). 
 *)
open Printf
open Cnf

type exp = (* evaluates to a Boolean *) 
  | Variable of Cnf.symbol
  | And of exp * exp 
  | Or of exp * exp
  | Implies of exp * exp 
  | Iff of exp * exp 
  | Not of exp 
  | Arithmetic of aexp 

and aexp = (* Evaluates to a Boolean *)
  | Le of anum * anum
  | Ge of anum * anum 
  | Eq of anum * anum 

and anum = (* Evaluates to a Number *)
  | Number of float 
  | Arithmetic_Variable of Cnf.symbol  
  | Plus of anum * anum 
  | Minus of anum * anum 
  | Times of anum * anum 

let rec exp_to_str exp = 
  match exp with
  | Variable(x) -> x
  | And(e1,e2) -> sprintf "(%s) && (%s)" (exp_to_str e1) (exp_to_str e2) 
  | Or(e1,e2) -> sprintf "(%s) || (%s)" (exp_to_str e1) (exp_to_str e2) 
  | Not(e1) -> sprintf "!(%s)" (exp_to_str e1) 
  | Implies(e1,e2) -> sprintf "(%s) -> (%s)" (exp_to_str e1) (exp_to_str e2) 
  | Iff(e1,e2) -> sprintf "(%s) <-> (%s)" (exp_to_str e1) (exp_to_str e2) 
  | Arithmetic(a) -> aexp_to_str a 
and aexp_to_str aexp =
  match aexp with
  | Le(a1,a2) -> sprintf "(%s) <= (%s)" (anum_to_str a1) (anum_to_str a2) 
  | Ge(a1,a2) -> sprintf "(%s) >= (%s)" (anum_to_str a1) (anum_to_str a2) 
  | Eq(a1,a2) -> sprintf "(%s) = (%s)" (anum_to_str a1) (anum_to_str a2) 
and anum_to_str anum =
  match anum with
  | Number(f) -> sprintf "%g" f 
  | Arithmetic_Variable(v) -> v 
  | Plus(a1,a2) -> sprintf "(%s) + (%s)" (anum_to_str a1) (anum_to_str a2) 
  | Minus(a1,a2) -> sprintf "(%s) - (%s)" (anum_to_str a1) (anum_to_str a2) 
  | Times(a1,a2) -> sprintf "(%s) * (%s)" (anum_to_str a1) (anum_to_str a2) 

(* We sometimes need to associate arithmetic expression (in our special
 * "theory") with SAT variables. We use a mapping to track such an
 * association. *) 
type mapping = (string * aexp) list 

module OrderedString =
struct
  type t = string
  let compare = compare
end
module StringSet = Set.Make(OrderedString) 
module StringMap = Map.Make(OrderedString) 

(*
 * Transform an arbitrary boolean expression into conjunctive normal form.
 *
 * This is useful because many SAT solvers require CNF input. 
 *
 * Subexpressions involving arithmetic cannot be converted to CNF. Instead,
 * each such subexpression is associated with a special new variable and
 * mapping is maintained to keep track of such associations.
 *
 * http://en.wikipedia.org/wiki/Conjunctive_normal_form#Converting_from_first-order_logic
 *)
let rec exp_to_cnf (exp : exp) : (cnf * mapping) = 

  let rec fixpoint convert exp =
    let exp' = convert exp in
    if exp' = exp then exp
    else fixpoint convert exp' 
  in

  (* Step 1 - remove Implies and IFF *) 
  let rec convert exp = 
    match exp with
    | Implies(a,b) -> Or(Not(convert a), convert b)
    | Iff(a,b) -> 
      let a = convert a in 
      let b = convert b in
      And(Or(a,Not(b)) , Or(Not(a), b))

    | And(a,b) -> And(convert a, convert b) 
    | Or(a,b) -> Or(convert a, convert b) 
    | Not(a) -> Not(convert a) 
    | x -> x
  in
  let exp = fixpoint convert exp in 

  (* Step 2 - push Not inward *)
  let rec convert exp =
    match exp with
    | Not(Or(p,q)) -> And(Not(convert p), Not(convert q))
    | Not(And(p,q)) -> Or(Not(convert p), Not(convert q))
    | Not(Not(p)) -> p 
    | Not(e) -> Not(convert e) 
    | And(a,b) -> And(convert a, convert b)
    | Or(a,b) -> Or(convert a, convert b)
    | x -> x
  in 
  let exp = fixpoint convert exp in

  (* Step 3 - distribute Or inward over And *)
  let rec convert exp =
    match exp with
    | Or(p, And(q,r))
    | Or(And(q,r), p) -> 
      let p = convert p in
      let q = convert q in
      let r = convert r in 
      And(Or(p,q), Or(p,r))
    | Not(e) -> Not(convert e)
    | And(a,b) -> And(convert a, convert b)
    | Or(a,b) -> Or(convert a, convert b)
    | x -> x
  in 
  let exp = fixpoint convert exp in 

  let fresh_variable_counter = ref 0 in 
  let new_variable () = 
    let var = sprintf "_%d" !fresh_variable_counter in
    incr fresh_variable_counter ;
    var
  in

  let rec exp_to_cnf exp : cnf * ((Cnf.symbol * aexp) list) =
    match exp with
    | Variable(sym) -> 
      let clause = [ (sym, true) ] in
      [ clause ], [] 
    | Not(Variable(sym)) -> 
      let clause = [ (sym, false) ] in
      [ clause ], [] 
    | And(e1, e2) ->
      let cnf1, mapping1 = exp_to_cnf e1 in 
      let cnf2, mapping2 = exp_to_cnf e2 in 
      cnf1 @ cnf2, mapping1 @ mapping2 
    | Or(_, _) -> begin 
      (* 
       * By the transformation above, subexpressions here can
       * only contain Or, Not or Variable (at the bare logic level)
       * or Theory expressions (such as Arithmetic). 
       *)
      let rec exp_to_clause exp : clause * mapping = 
        match exp with 
        | Variable(sym) -> [(sym, true)], [] 
        | Not(Variable(sym)) -> [(sym, false)], [] 
        | Or(e1,e2) -> 
          let c1, m1 = exp_to_clause e1 in 
          let c2, m2 = exp_to_clause e2 in 
          c1 @ c2, m1 @ m2 
        | Arithmetic(a) -> 
          let v = new_variable () in 
          [(v, true)], [(v, a)]
        | Not(Arithmetic(a)) -> 
          let v = new_variable () in 
          [(v, false)], [(v, a)]
        | _ -> failwith "exp_to_clause: impossible (1)" 
      in
      let clause, mapping = exp_to_clause exp in
      [ clause ], mapping
    end 
    | Arithmetic(a) -> 
      let v = new_variable () in
      let clause = [ (v, true) ] in
      [ clause ], [(v, a)] 
    | Not(Arithmetic(a)) -> 
      let v = new_variable () in
      let clause = [ (v, false) ] in
      [ clause ], [(v, a)] 
    | _ -> failwith "exp_to_cnf: impossible (2)" 
  in
  exp_to_cnf exp 

File: solver/exp.hs
module Exp (Exp(..),
            AExp(..),
            Anum(..),
            convertToConjuctiveNormalForm) where

    import Control.Monad.Fix

    data Exp =
        Variable String |
        And Exp Exp |
        Or Exp Exp |
        Implies Exp Exp |
        Iff Exp Exp | 
        Not Exp |
        Arithmetic AExp
        deriving (Eq)

    data AExp =
        Le Anum Anum |
        Ge Anum Anum |
        Eq Anum Anum
        deriving (Eq)

    data Anum = 
        Constant Float |
        ArithmeticVariable String |
        Add Anum Anum |
        Minus Anum Anum |
        Times Anum Anum 
        deriving (Eq)
                                  
    eliminateIffAndImplies :: Exp -> Exp
    eliminateIffAndImplies (Iff a b) =
        let a1 = (eliminateIffAndImplies a) in
        let b1 = (eliminateIffAndImplies b) in
        (And (Or (Not a1) b1) (Or (Not b1) a1))
    eliminateIffAndImplies e@(Variable _) = e
    eliminateIffAndImplies (And e1 e2) =
        And (eliminateIffAndImplies e1)
            (eliminateIffAndImplies e2)
    eliminateIffAndImplies (Or e1 e2) =
        Or (eliminateIffAndImplies e1)
           (eliminateIffAndImplies e2)
    eliminateIffAndImplies (Not e) =
        Not (eliminateIffAndImplies e)
    eliminateIffAndImplies (Implies a b) =
        Or (Not (eliminateIffAndImplies a)) (eliminateIffAndImplies b)
    eliminateIffAndImplies e@(Arithmetic _) = e 

    -- assumes no => or <=> in the expression
    applyDeMorgan :: Exp -> Exp
    applyDeMorgan e@(Variable var) = e
    applyDeMorgan (And a b) = And (applyDeMorgan a) (applyDeMorgan b)
    applyDeMorgan (Or a b) = Or (applyDeMorgan a) (applyDeMorgan b)
    applyDeMorgan (Not e) =
        case e of
          Or a b -> And (Not (applyDeMorgan a)) (Not (applyDeMorgan b))
          And a b -> Or (Not (applyDeMorgan a)) (Not (applyDeMorgan b))
          e -> Not (applyDeMorgan e)
    applyDeMorgan e@(Arithmetic _) = e
    applyDeMorgan (Implies _ _ ) = undefined
    applyDeMorgan (Iff _ _) = undefined

    -- assumes no => or <=> in the expression
    -- assumes negations olny in front of literals or
    -- other negations (i.e. we eliminated iff and implies
    -- and we applied DeMorgan laws)
    eliminateDoubleNegation :: Exp -> Exp
    eliminateDoubleNegation e@(Variable _) = e
    eliminateDoubleNegation (And a b) = And (eliminateDoubleNegation a) (eliminateDoubleNegation b)
    eliminateDoubleNegation (Or a b) = Or (eliminateDoubleNegation a) (eliminateDoubleNegation b)
    eliminateDoubleNegation (Implies _ _) = undefined
    eliminateDoubleNegation (Iff _ _) = undefined
    eliminateDoubleNegation (Not e) = case e of
                                        Not e1 -> e1
                                        _ -> Not (eliminateDoubleNegation e)
    eliminateDoubleNegation e@(Arithmetic _) = e

    
    eliminateConjInDisj :: Exp -> Exp
    eliminateConjInDisj e@(Variable _) = e
    eliminateConjInDisj (And e1 e2) = And (eliminateConjInDisj e1) (eliminateConjInDisj e2)
    eliminateConjInDisj (Or (And a b) c) =
        let a' = eliminateConjInDisj a in
        let b' = eliminateConjInDisj b in
        let c' = eliminateConjInDisj c in
        (And (Or a' c') (Or b' c'))
    eliminateConjInDisj (Or a (And b c)) =
        let a' = eliminateConjInDisj a in
        let b' = eliminateConjInDisj b in
        let c' = eliminateConjInDisj c in
        (And (Or a' b') (Or a' c'))
    eliminateConjInDisj (Or a b) = Or (eliminateConjInDisj a) (eliminateConjInDisj b)
    eliminateConjInDisj (Implies _ _ ) = undefined
    eliminateConjInDisj (Iff _ _) = undefined
    eliminateConjInDisj (Not e) = Not (eliminateConjInDisj e)
    eliminateConjInDisj e@(Arithmetic _) = e
          
    fixed :: (Exp ->Exp) -> Exp -> Exp
    fixed f e = let e1 = f e in
                if e1 == e then e1
                else fixed f e1

    --expression to Conjuctive Normal Form
    convertToConjuctiveNormalForm :: Exp -> Exp
    convertToConjuctiveNormalForm e =
        let e1 = eliminateIffAndImplies e in
        let e2 = fixed applyDeMorgan e1 in
        let e3 = fixed eliminateDoubleNegation e2 in
        fixed eliminateConjInDisj e3

    instance Show(Exp) where
           show (Variable var) = var
           show (And e1 e2) = "("++(show e1)++") && ("
                            ++ (show e2) ++ ")"
           show (Or e1 e2) = "("++(show e1)++") || ("
                            ++ (show e2) ++ ")"
           show (Implies e1 e2) = "("++(show e1)++") => ("
                            ++ (show e2) ++ ")"
           show (Iff e1 e2) = "("++(show e1)++") <=> ("
                            ++ (show e2) ++ ")"
           show (Not e) = "!("++ (show e) ++")"
           show (Arithmetic aexp) = show aexp


    instance Show(AExp) where
           show (Le anum1 anum2) = "("++ (show anum1) ++ ")<("
                                   ++ (show anum2) ++ ")"
           show (Ge anum1 anum2) = "("++ (show anum1) ++ ")>("
                                   ++ (show anum2) ++ ")"
           show (Eq anum1 anum2) = "("++ (show anum1) ++ ")==("
                                   ++ (show anum2) ++ ")"

    instance Show(Anum) where
           show (Constant ct) = show ct
           show (ArithmeticVariable var) = var
           show (Add anum1 anum2) = "(" ++ (show anum1) ++ ")+("
                                    ++ (show anum2) ++ ")"
           show (Minus anum1 anum2) = "(" ++ (show anum1) ++ ")-("
                                    ++ (show anum2) ++ ")"
           show (Times anum1 anum2) = "(" ++ (show anum1) ++ ")*("
                                    ++ (show anum2) ++ ")"
